eax/src/online.rs

```diff
@@ -0,0 +1,326 @@
+        // 1. n ← OMAC(0 || Nonce)
+        // (the 0 means the number zero in l bits)
+        let n = prepend_cmac(&key, 0, nonce);
+        let n = n.finalize().into_bytes();
+
+        // NOTE: These can be updated online later
+        // 2. h ← OMAC(1 || associated data)
+        let h = prepend_cmac(&key, 1, &[]);
+        // 3. c ← OMAC(2 || enc)
+        let c = prepend_cmac(&key, 2, &[]);
+
+        let cipher = ctr::Ctr128::<Cipher>::from_block_cipher(Cipher::new(&key), &n);
+
+        Eax {
+            nonce: n,
+            data: h,
+            message: c,
+            ctr: cipher,
+            marker: PhantomData,
+            bomb: DropBomb,
+        }
+    }
+}
+
+impl<Cipher, Op> Eax<Cipher, Op>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+    Op: CipherOp,
+{
+    /// Process the associated data (AD).
+    #[inline]
+    pub fn update_assoc(&mut self, aad: &[u8]) {
+        self.data.update(aad);
+    }
+
+    /// Derives the tag from the encrypted/decrypted message so far.
+    ///
+    /// NOTE: This has to be called when the value is consumed.
+    #[inline]
+    fn finish_inner(self) -> Tag {
+        self.bomb.defuse();
+
+        let h = self.data.finalize().into_bytes();
+        let c = self.message.finalize().into_bytes();
+
+        self.nonce.zip(h, |a, b| a ^ b).zip(c, |a, b| a ^ b)
+    }
+
+    /// Derives the tag from the encrypted/decrypted message so far.
+    ///
+    /// If the encryption/decryption operation is finished, [`finish`] method
+    /// *must* be called instead.
+    ///
+    ///[`finish`]: #method.finish
+    #[inline]
+    pub fn tag_clone(&self) -> Tag {
+        let h = self.data.clone().finalize().into_bytes();
+        let c = self.message.clone().finalize().into_bytes();
+
+        self.nonce.zip(h, |a, b| a ^ b).zip(c, |a, b| a ^ b)
+    }
+}
+
+impl<Cipher> Eax<Cipher, Encrypt>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+{
+    /// Applies encryption to the plaintext.
+    #[inline]
+    pub fn encrypt(&mut self, msg: &mut [u8]) {
+        self.ctr.apply_keystream(msg);
+        self.message.update(msg);
+    }
+
+    /// Finishes the encryption stream, returning the derived tag.
+    ///
+    /// This *must* be called after the stream encryption is finished.
+    #[must_use = "tag must be saved to later verify decrypted data"]
+    #[inline]
+    pub fn finish(self) -> Tag {
+        self.finish_inner()
+    }
+}
+
+impl<Cipher> Eax<Cipher, Decrypt>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+{
+    /// Applies decryption to the ciphertext.
+    #[inline]
+    pub fn decrypt(&mut self, msg: &mut [u8]) {
+        self.message.update(msg);
+        self.ctr.apply_keystream(msg);
+    }
```

Comment:
Like @newpavlov mentioned above, at the very least this needs to be market as hazmat (e.g. `decrypt_unauthenticated_hazmat` with a comment explaining how potential misuse can lead to catastrophic failures like chosen ciphertext attacks and potentially full plaintext recovery)